# Data preprocessing

In [1]:
import numpy as np
import pandas as pd
import json
import matplotlib.pyplot as plt

In [2]:
def read_json(file):
    reader = pd.read_json(f"Data/{file}.json")
    return reader[['onetsoc_code','element_id', 'scale_id', 'data_value']]

In [3]:
knowledge = read_json("knowledge")
abilities = read_json("abilities")
skills = read_json("skills")
work_activities = read_json("work_activities")

In [4]:
abilities.head(10)

onetsoc_code element_id scale_id  data_value
0   11-1011.00  1.A.1.a.1       IM        4.50
1   11-1011.00  1.A.1.a.1       LV        4.88
2   11-1011.00  1.A.1.a.2       IM        4.25
3   11-1011.00  1.A.1.a.2       LV        4.62
4   11-1011.00  1.A.1.a.3       IM        4.38
5   11-1011.00  1.A.1.a.3       LV        5.00
6   11-1011.00  1.A.1.a.4       IM        4.12
7   11-1011.00  1.A.1.a.4       LV        4.62
8   11-1011.00  1.A.1.b.1       IM        3.88
9   11-1011.00  1.A.1.b.1       LV        4.62

In [5]:
knowledge.head(10)

onetsoc_code element_id scale_id  data_value
0   11-1011.00    2.C.1.a       IM        4.75
1   11-1011.00    2.C.1.a       LV        6.23
2   11-1011.00    2.C.1.b       IM        2.66
3   11-1011.00    2.C.1.b       LV        3.50
4   11-1011.00    2.C.1.c       IM        3.70
5   11-1011.00    2.C.1.c       LV        4.36
6   11-1011.00    2.C.1.d       IM        3.23
7   11-1011.00    2.C.1.d       LV        3.90
8   11-1011.00    2.C.1.e       IM        4.09
9   11-1011.00    2.C.1.e       LV        5.55

In [6]:
skills.tail(10)

onetsoc_code element_id scale_id  data_value
67008   53-7121.00    2.B.4.h       IM        2.12
67009   53-7121.00    2.B.4.h       LV        2.12
67010   53-7121.00    2.B.5.a       IM        3.12
67011   53-7121.00    2.B.5.a       LV        2.88
67012   53-7121.00    2.B.5.b       IM        2.00
67013   53-7121.00    2.B.5.b       LV        1.12
67014   53-7121.00    2.B.5.c       IM        2.00
67015   53-7121.00    2.B.5.c       LV        1.88
67016   53-7121.00    2.B.5.d       IM        2.88
67017   53-7121.00    2.B.5.d       LV        2.75

In [7]:
work_activities.tail(10)

onetsoc_code element_id scale_id  data_value
79320   53-7121.00  4.A.4.b.5       IM        3.07
79321   53-7121.00  4.A.4.b.5       LV        3.28
79322   53-7121.00  4.A.4.b.6       IM        2.72
79323   53-7121.00  4.A.4.b.6       LV        2.40
79324   53-7121.00  4.A.4.c.1       IM        2.54
79325   53-7121.00  4.A.4.c.1       LV        2.27
79326   53-7121.00  4.A.4.c.2       IM        1.93
79327   53-7121.00  4.A.4.c.2       LV        1.60
79328   53-7121.00  4.A.4.c.3       IM        2.56
79329   53-7121.00  4.A.4.c.3       LV        2.64

### Abilities

In [8]:
len(abilities["onetsoc_code"].value_counts())

962

In [9]:
abilities["onetsoc_code"].value_counts().sort_values()

17-2111.01     31
17-2131.00     56
33-1021.02    103
43-4041.02    104
11-9021.00    104
             ... 
51-4072.00    104
31-9095.00    104
51-9198.00    104
53-6011.00    104
29-9092.00    104
Name: onetsoc_code, Length: 962, dtype: int64

In [10]:
len(abilities.groupby(["element_id"]).size())

52

### Required knowledge

In [11]:
len(knowledge["onetsoc_code"].value_counts())

967

In [12]:
knowledge["onetsoc_code"].value_counts().sort_values()

43-4111.00    55
43-4131.00    61
53-1021.01    66
31-1013.00    66
35-9011.00    66
              ..
51-9198.00    66
51-9193.00    66
53-2021.00    66
39-5011.00    66
51-6062.00    66
Name: onetsoc_code, Length: 967, dtype: int64

In [13]:
len(knowledge.groupby(["element_id"]).size())

33

### Required skills

In [14]:
len(skills["onetsoc_code"].value_counts())

959

In [15]:
skills["onetsoc_code"].value_counts().sort_values()

49-2093.00    29
33-9021.00    32
49-3021.00    42
33-3051.01    65
43-4071.00    70
              ..
51-9198.00    70
51-9193.00    70
53-2021.00    70
33-3021.03    70
53-2012.00    70
Name: onetsoc_code, Length: 959, dtype: int64

In [16]:
len(skills.groupby(["element_id"]).size())

35

### Work activities

In [17]:
len(work_activities["onetsoc_code"].value_counts())

968

In [18]:
work_activities["onetsoc_code"].value_counts().sort_values()

47-1011.00    36
31-1013.00    82
35-9011.00    82
37-3013.00    82
49-3052.00    82
              ..
51-9193.00    82
53-2021.00    82
29-1067.00    82
51-9032.00    82
51-6062.00    82
Name: onetsoc_code, Length: 968, dtype: int64

In [19]:
len(work_activities.groupby(["element_id"]).size())

41

## Reducing number of jobs

In [20]:
c = 0

for job in skills["onetsoc_code"].unique():
    if job in list(knowledge["onetsoc_code"]) :
        c+=1  
    else:
        print("In knowledge dataset there is no any data about: " + str(job) + " job")
print("There is " + str(c) + " matches between jobs described in skills and knowledge dataset")

In knowledge dataset there is no any data about: 43-4121.00 job
There is 958 matches between jobs described in skills and knowledge dataset


In [21]:
c = 0

for job in skills["onetsoc_code"].unique():
    if job in list(work_activities["onetsoc_code"]) :
        c+=1  
    else:
        print("In knowledge dataset there is no any data about: " + str(job) + " job")
print("There is " + str(c) + " matches between jobs described in skills and work_activities dataset")

There is 959 matches between jobs described in skills and work_activities dataset


In [22]:
c = 0

for job in skills["onetsoc_code"].unique():
    if job in list(abilities["onetsoc_code"]) :
        c+=1  
    else:
        print("In abilities dataset there is no any data about: " + str(job) + " job")
print("There is " + str(c) + " matches between jobs described in skills and abilities dataset")

In abilities dataset there is no any data about: 17-2111.02 job
In abilities dataset there is no any data about: 17-2111.03 job
In abilities dataset there is no any data about: 17-2112.00 job
In abilities dataset there is no any data about: 17-2112.01 job
In abilities dataset there is no any data about: 17-2121.01 job
In abilities dataset there is no any data about: 17-2121.02 job
There is 953 matches between jobs described in skills and abilities dataset


## From the analysis above we can conclude that:

- All of three datasets should have 958 job titles (We need to drop "43-4121.00" because knowledge dataset does not contain any data for it)
- <b>Skills dataset</b> does not contain 70 skills for: <b>49-2093.00, 33-9021.00, 49-3021.00, 33-3051.01</b> as others. It means that they would be dropped.
- The same thing is happening for job titles: <b>43-4111.00, 43-4131.00 in knowledge dataset</b>.
- <b>47-1011.00 would be dropped from work_activities because of that reason</b>
- <b>17-2111.01, 17-2131.00, 33-1021.02 would be dropped from the abilities dataset</b>

# Let's start with cleaning


### Step 1:

In [23]:
knowledge.drop(knowledge.loc[knowledge['onetsoc_code']=="43-4121.00"].index, inplace=True)

In [24]:
skills.drop(skills.loc[skills['onetsoc_code']=="43-4121.00"].index, inplace=True)

### Step 2:

In [25]:
skills.drop(skills.loc[skills['onetsoc_code']=="49-2093.00"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="33-9021.00"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="49-3021.00"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="33-3051.01"].index, inplace=True)

#because of abilities
skills.drop(skills.loc[skills['onetsoc_code']=="17-2111.02"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2111.03"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2112.00"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2112.01"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2121.01"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2121.02"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="33-1021.02"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2131.00"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="17-2111.010"].index, inplace=True)

### Step 3:

In [26]:
knowledge.drop(knowledge.loc[knowledge['onetsoc_code']=="43-4111.00"].index, inplace=True)
knowledge.drop(knowledge.loc[knowledge['onetsoc_code']=="43-4131.00"].index, inplace=True)

skills.drop(skills.loc[skills['onetsoc_code']=="43-4111.00"].index, inplace=True)
skills.drop(skills.loc[skills['onetsoc_code']=="43-4131.00"].index, inplace=True)

work_activities.drop(work_activities.loc[work_activities['onetsoc_code']=="43-4111.00"].index, inplace=True)
work_activities.drop(work_activities.loc[work_activities['onetsoc_code']=="43-4131.00"].index, inplace=True)

In [27]:
work_activities.drop(work_activities.loc[work_activities['onetsoc_code']=="47-1011.00"].index, inplace=True)

knowledge.drop(knowledge.loc[knowledge['onetsoc_code']=="47-1011.00"].index, inplace=True)

skills.drop(skills.loc[skills['onetsoc_code']=="47-1011.00"].index, inplace=True)

### Step 4:

In [28]:
abilities.drop(abilities.loc[abilities['onetsoc_code']=="17-2111.010"].index, inplace=True)
abilities.drop(abilities.loc[abilities['onetsoc_code']=="17-2131.00"].index, inplace=True)
abilities.drop(abilities.loc[abilities['onetsoc_code']=="33-1021.02"].index, inplace=True)

# Concatenating datasets

In [29]:
knowledge = knowledge[knowledge["onetsoc_code"].isin(skills["onetsoc_code"])]

In [30]:
work_activities = work_activities[work_activities["onetsoc_code"].isin(skills["onetsoc_code"])]

In [31]:
abilities = abilities[abilities["onetsoc_code"].isin(skills["onetsoc_code"])]

In [32]:
print("Work activities shape: " + str(work_activities.shape) + "\n" + "Skills shape: " + str(skills.shape) + "\n" + "Knowledge shape: " + str(knowledge.shape) + "\n" + "Abilities shape: " + str(abilities.shape))

Work activities shape: (77326, 4)
Skills shape: (66010, 4)
Knowledge shape: (62238, 4)
Abilities shape: (97999, 4)


### Computing mean of skills

In [33]:
knowledge = knowledge.groupby(["onetsoc_code", "element_id"]).mean().reset_index()
skills = skills.groupby(["onetsoc_code", "element_id"]).mean().reset_index()
work_activities = work_activities.groupby(["onetsoc_code", "element_id"]).mean().reset_index()
abilities = abilities.groupby(["onetsoc_code", "element_id"]).mean().reset_index()

In [34]:
abilities.head()

onetsoc_code element_id  data_value
0   11-1011.00  1.A.1.a.1       4.690
1   11-1011.00  1.A.1.a.2       4.435
2   11-1011.00  1.A.1.a.3       4.690
3   11-1011.00  1.A.1.a.4       4.370
4   11-1011.00  1.A.1.b.1       4.250

In [35]:
knowledge.head()

onetsoc_code element_id  data_value
0   11-1011.00    2.C.1.a       5.490
1   11-1011.00    2.C.1.b       3.080
2   11-1011.00    2.C.1.c       4.030
3   11-1011.00    2.C.1.d       3.565
4   11-1011.00    2.C.1.e       4.820

In [36]:
skills.shape

(33005, 3)

In [37]:
work_activities.head()

onetsoc_code element_id  data_value
0   11-1011.00  4.A.1.a.1       5.035
1   11-1011.00  4.A.1.a.2       4.220
2   11-1011.00  4.A.1.b.1       4.555
3   11-1011.00  4.A.1.b.2       2.245
4   11-1011.00  4.A.1.b.3       3.655

In [38]:
skills = skills.pivot(index='onetsoc_code', columns='element_id', values='data_value')
knowledge = knowledge.pivot(index='onetsoc_code', columns='element_id', values='data_value')
work_activities = work_activities.pivot(index='onetsoc_code', columns='element_id', values='data_value')
abilities = abilities.pivot(index='onetsoc_code', columns='element_id', values='data_value')

In [39]:
final_dataset = pd.concat([skills, knowledge, work_activities, abilities], axis = 1)

In [2]:
final_dataset.shape

NameError: name 'final_dataset' is not defined

In [41]:
final_dataset.head()

element_id    2.A.1.a  2.A.1.b  2.A.1.c  2.A.1.d  2.A.1.e  2.A.1.f  2.A.2.a  \
onetsoc_code                                                                  
11-1011.00      4.435     4.50    4.190    4.630    3.435    1.500    4.565   
11-1011.03      4.125     4.00    4.130    4.125    3.065    1.625    4.060   
11-1021.00      3.940     4.00    3.565    4.000    2.440    1.500    3.940   
11-2011.00      3.875     4.12    3.815    4.060    3.125    1.120    4.060   
11-2021.00      4.065     4.00    3.565    4.000    2.935    1.625    4.065   

element_id    2.A.2.b  2.A.2.c  2.A.2.d  ...  1.A.4.a.3  1.A.4.a.4  1.A.4.a.5  \
onetsoc_code                             ...                                    
11-1011.00      4.375    3.500     4.81  ...      1.750       0.50       0.50   
11-1011.03      3.625    3.440     3.87  ...      1.810       0.50       0.50   
11-1021.00      3.560    3.125     4.00  ...      2.125       0.94       1.19   
11-2011.00      3.685    3.125     3.75  ...      2.815       0.50       0.50   
11-2021.00      4.000    3.310     4.00  ...      2.880       0.50       0.50   

element_id    1.A.4.a.6  1.A.4.a.7  1.A.4.b.1  1.A.4.b.2  1.A.4.b.3  \
onetsoc_code                                                          
11-1011.00        1.625       0.50      1.625       2.12       0.50   
11-1011.03        1.940       0.50      1.750       1.75       0.62   
11-1021.00        1.940       0.88      2.000       2.06       1.06   
11-2011.00        1.375       0.50      1.500       1.50       0.50   
11-2021.00        1.250       0.50      1.690       1.75       0.50   

element_id    1.A.4.b.4  1.A.4.b.5  
onetsoc_code                        
11-1011.00        4.370      4.500  
11-1011.03        3.810      3.810  
11-1021.00        3.565      3.625  
11-2011.00        3.940      3.940  
11-2021.00        3.750      3.815  

[5 rows x 161 columns]

In [42]:
final_dataset.to_csv('mean_dataset.csv')